### ISSUE TITLE
How to properly close shutdown an apollo client version 3?

### ISSUE THREAD

**Issue by AddictArts** [apollographql/apollo-kotlin#3802]
Please search the [existing Github issues](https://github.com/apollographql/apollo-android/issues), [Community Discourse](https://community.apollographql.com/tag/mobile), [StackOverflow](https://stackoverflow.com/questions/tagged/apollo-android) before asking your question. Questions are welcome but avoid duplicating issues and discussions. If you'd like a Github issue to be reopen, please leave a message on the issue.

For an informal chat about the project, you can also [say hi in the #apollo-android channel in the KotlinLang Slack](https://app.slack.com/client/T09229ZC6/C01A6KM1SBZ)(get your invite [here](https://slack.kotl.in/)).

**Question.**
Enter your question here.
With version 3 a simple fun main with a runBlocking calling a query execute will continue to run and not exit. When commenting out the Apollo code the main exits and process exits. Thanks for the help.

**Comment by martinbonnin** [comment]
Hi 👋 

You can use `ApolloClient.dispose()` to free resources associated with the `ApolloClient`. Note that this doesn't shut down the okHttpClient as it may be shared with other users. If you want to do so, keep a reference to the `OkHttpClient` and call `okHttpClient.dispatcher.executorService.shutdown()` as you indicated above (the difference in methods vs properties is for okhttp 3 vs okhttp 4)

**Comment by martinbonnin** [comment]
My bad, looks like there's a regression there that keeps a thread open. Will keep looking and update this thread.

**Comment by martinbonnin** [comment]
Indeed, this should be fixed with https://github.com/apollographql/apollo-kotlin/pull/3803. Sorry about this. Until next version, you can use the following workaround:

```kotlin
      ApolloClient.Builder()
      .serverUrl("https://www.google.com")
      .okHttpClient(okHttpClient)
      .subscriptionNetworkTransport(object: NetworkTransport {
        override fun dispose() {
        }
        override fun <D : Operation.Data> execute(request: ApolloRequest<D>): Flow<ApolloResponse<D>> {
          TODO("Not yet implemented")
        }
      })
```

**Comment by AddictArts** [comment]
Thanks  for validating. That led to
`Caused by: java.lang.IllegalStateException: Apollo: 'webSocketEngine' has no effect if 'subscriptionNetworkTransport' is set
	at com.apollographql.apollo3.ApolloClient$Builder.build(ApolloClient.kt:482)`

**Comment by martinbonnin** [comment]
ooops, my bad.`okHttpClient` will indeed set a webSocketEngine based on `OkHttp` too. Use `httpEngine` instead:

```kotlin
  val apolloClient = ApolloClient.Builder()
    .serverUrl("https://www.google.com")
    .httpEngine(DefaultHttpEngine(okHttpClient))
    .subscriptionNetworkTransport(object : NetworkTransport {
      override fun dispose() {
      }
      override fun <D : Operation.Data> execute(request: ApolloRequest<D>): Flow<ApolloResponse<D>> {
        TODO("Not yet implemented")
      }
    })
    .build()
```
